File: geekbang/algorithm-models/7.2- Dijkstra 求最短路代码模板/dijkstra.cpp
// C/C++
// https://www.acwing.com/problem/content/description/852/
#include<bits/stdc++.h>
using namespace std;
const int MAX_N = 150005, MAX_M = 150005;
vector<int> ver[MAX_N]; // 出边数组 - 另一端点
vector<int> edge[MAX_N]; // 出边数组 - 边权
int n, m, d[MAX_N];
bool v[MAX_N];
// pair<-dist[x], x>
priority_queue<pair<int, int>> q;


// 插入一条从x到y长度z的有向边
void add(int x, int y, int z) {
    ver[x].push_back(y);
    edge[x].push_back(z);
}


int main() {
    cin >> n >> m;
    for (int i = 1; i <= m; i++) {
        int x, y, z;
        scanf("%d%d%d", &x, &y, &z);
        add(x, y, z);
    }
    memset(d, 0x7f, sizeof(d));
    d[1] = 0;
    q.push(make_pair(0, 1));
    while (!q.empty()) {
        int x = q.top().second;
        q.pop();
        if (v[x]) continue;
        v[x] = true;
        for (int i = 0; i < ver[x].size(); i++) {
            int y = ver[x][i], z = edge[x][i];
            if (d[y] > d[x] + z) {
                d[y] = d[x] + z;
                q.push(make_pair(-d[y], y));
            }
        }
    }
    if (d[n] == 0x7f7f7f7f) puts("-1");
    else cout << d[n] << endl;
}

File: geekbang/algorithm-models/7.2- Dijkstra 求最短路代码模板/dijkstra.go
// Go
// https://www.acwing.com/problem/content/description/852/

package main

import (
	"fmt"
)

type Pair struct {
	A, B int
}

type PriorityQueue struct {
	Data []Pair
}

func (q *PriorityQueue) Add(p Pair) {
	q.Data = append(q.Data, p)
	cur := len(q.Data) - 1
	for {
		if cur != 0 && q.Data[cur].A < q.Data[(cur-1)/2].A {
			q.Data[cur], q.Data[(cur-1)/2] = q.Data[(cur-1)/2], q.Data[cur]
			cur = (cur - 1) / 2
		} else {
			return
		}
	}
}

func (q *PriorityQueue) Pop() (res Pair) {
	res = q.Data[0]
	q.Data[0] = q.Data[len(q.Data)-1]
	q.Data = q.Data[:len(q.Data)-1]
	cur := 0
	for {
		l, r := cur*2+1, cur*2+2
		if l < len(q.Data) && q.Data[l].A < q.Data[cur].A && (r >= len(q.Data) || q.Data[l].A <= q.Data[r].A) {
			q.Data[cur], q.Data[l] = q.Data[l], q.Data[cur]
			cur = l
		} else if r < len(q.Data) && q.Data[r].A < q.Data[cur].A && (l >= len(q.Data) || q.Data[r].A <= q.Data[l].A) {
			q.Data[cur], q.Data[r] = q.Data[r], q.Data[cur]
			cur = r
		} else {
			return
		}
	}
}

const maxN = 150005

type Edge struct {
	V, W, Nxt int
}

var e = make([]Edge, maxN)
var head = make([]int, maxN)
var cnt = 1

func AddEdge(u, v, w int) {
	e[cnt].V = v
	e[cnt].W = w
	e[cnt].Nxt = head[u]
	head[u] = cnt
	cnt++
}

var n, m int
var dis = make([]int, maxN)
var q = PriorityQueue{Data: make([]Pair, 0)}

func main() {
	for i := range dis {
		dis[i] = -1
	}
	fmt.Scanf("%d %d", &n, &m)
	for ; m > 0; m-- {
		var u, v, w int
		fmt.Scanf("%d %d %d", &u, &v, &w)
		AddEdge(u, v, w)
	}
	dis[1] = 0
	q.Add(Pair{A: 0, B: 1})
	for len(q.Data) > 0 {
		cur := q.Pop()
		if cur.A != dis[cur.B] {
			continue
		} else if cur.B == n {
			break
		}
		for j := head[cur.B]; j != 0; j = e[j].Nxt {
			if dis[e[j].V] == -1 || cur.A+e[j].W < dis[e[j].V] {
				dis[e[j].V] = cur.A + e[j].W
				q.Add(Pair{A: dis[e[j].V], B: e[j].V})
			}
		}
	}
	fmt.Println(dis[n])
}
